src/graph/sampler.cc

```diff
@@ -0,0 +1,462 @@
+        // If the sampled neighbor is inserted to the map successfully.
+        if (ret.second) {
+          sub_vers.emplace_back(tmp_sampled_src_list[i], cur_node_level + 1);
+        }
+      }
+    }
+    layer_offsets[layer_id + 1] = layer_offsets[layer_id] + sub_ver_map.size();
+  }
+
+  uint64_t num_vertices = sub_vers.size();
+  NodeFlow nf;
+  nf.node_mapping = IdArray::Empty({static_cast<int64_t>(num_vertices)},
+                                   DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  nf.edge_mapping = IdArray::Empty({static_cast<int64_t>(num_edges)},
+                                   DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  nf.layer_offsets = IdArray::Empty({static_cast<int64_t>(num_hops + 1)},
+                                    DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  nf.flow_offsets = IdArray::Empty({static_cast<int64_t>(num_hops)},
+                                    DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+
+  dgl_id_t *out = static_cast<dgl_id_t *>(nf.node_mapping->data);
+  dgl_id_t *out_layer = static_cast<dgl_id_t *>(nf.layer_offsets->data);
+  dgl_id_t *out_flow = static_cast<dgl_id_t *>(nf.flow_offsets->data);
+  dgl_id_t* val_list_out = static_cast<dgl_id_t *>(nf.edge_mapping->data);
+
+  // Construct sub_csr_graph
+  auto subg_csr = std::make_shared<CSR>(num_vertices, num_edges);
+  subg_csr->indices.resize(num_edges);
+  subg_csr->edge_ids.resize(num_edges);
+  dgl_id_t* col_list_out = subg_csr->indices.data();
+  int64_t* indptr_out = subg_csr->indptr.data();
+  size_t collected_nedges = 0;
+
+  // The data from the previous steps:
+  // * node data: sub_vers (vid, layer), neigh_pos,
+  // * edge data: neighbor_list, probability.
+  // * layer_offsets: the offset in sub_vers.
+  dgl_id_t ver_id = 0;
+  std::vector<std::unordered_map<dgl_id_t, dgl_id_t>> layer_ver_maps;
+  layer_ver_maps.resize(num_hops);
+  size_t out_node_idx = 0;
+  for (int layer_id = num_hops - 1; layer_id >= 0; layer_id--) {
+    // We sort the vertices in a layer so that we don't need to sort the neighbor Ids
+    // after remap to a subgraph.
+    std::sort(sub_vers.begin() + layer_offsets[layer_id],
+              sub_vers.begin() + layer_offsets[layer_id + 1],
+              [](const std::pair<dgl_id_t, dgl_id_t> &a1,
+                 const std::pair<dgl_id_t, dgl_id_t> &a2) {
+      return a1.first < a2.first;
+    });
+
+    // Save the sampled vertices and its layer Id.
+    for (size_t i = layer_offsets[layer_id]; i < layer_offsets[layer_id + 1]; i++) {
+      out[out_node_idx++] = sub_vers[i].first;
+      layer_ver_maps[layer_id].insert(std::pair<dgl_id_t, dgl_id_t>(sub_vers[i].first, ver_id++));
+      assert(sub_vers[i].second == layer_id);
+    }
+  }
+  CHECK(out_node_idx == num_vertices);
+
+  // Copy sampled results to a CSR.
+  // When we copy data, we need to move the last layer to the first layer
+  // because we consider the layer with the leaf nodes as layer 0.
```

Comment:
Any particular reason for reversing the direction?

Reply:
sampling algorithms have to start from the seed nodes, so the seed nodes are in the first layer and the input nodes are in the last layer. When we expose the sampled graph to a Python user, we say the input nodes are in the first layer and the seed nodes are in the last layer.

Reply:
I see.  Maybe replace the comment with your reply?